### PR TITLE
allow unicode flat and sharp as input for chord symbols

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
also double flat and double sharp, &Delta;, &deg;, &oslash; and &Oslash;. Comes up at times. last time in https://musescore.org/en/node/269352

### CODE REVIEW

**Comment by MarcSabatella** [comment]
I haven't built this to test, but it kind of surprises me that it works to just change this here.  Seems that will only catch accidentals used in the chord root (or bass note) but not in the chord extension.

**Comment by Jojo-Schmitz** [comment]
I've toyed with this some 2 years ago, so indeed would need to test this (again?).

**Comment by Jojo-Schmitz** [comment]
I've toyed with this some 2 years ago, so indeed would need to test this (again?).

Doesn't work at all currently, the whole chord symbol rendering seems broken or disabled in master.

**Comment by Jojo-Schmitz** [comment]
Changed it to replace them all on input already

**Comment by MarcSabatella** [comment]
I don't see how that would work either - as far as I can tell the replacement happens before the text exists.  Have you verified this works?

**Comment by Jojo-Schmitz** [comment]
No, not yet. It is carnival here ;-)

**Comment by MarcSabatella** [comment]
OK, I finally tried it, and no, it does not work.  To do the replacement, you'd need to treat the keystrokes as they come in (in edit) or when finished (in endEdit) or when parsed (in parseHarmony).  A quick test suggests doing it at the very beginning of Harmony::endEdit() seems to work.  More testing would still be required to make sure it really works correctly with respect to parts, transposition, etc.

**Comment by Jojo-Schmitz** [comment]
Thanks, I'll move them there, tomorrow ;-)

**Comment by MarcSabatella** [comment]
BTW, another approach is to allow the accidentals but parse them normally.  That is, add them to the parser in chordlist.cpp (ParsedChord::configure) as well as convertNote().  This approach would mean they stay Unicode even when the user goes back to edit the chord later.  The replace approach means if he edits later, he'll see "b" et al.  Maybe that's better, then he'll learn he doesn't need the Unicode characters.

**Comment by Jojo-Schmitz** [comment]
How exactly did you test it in `Harmony::endEdit()`? Wouldn't it need to be done in `Harmony::edit()`, just before the call to `Harmony::parseHarmony()`, for that not to fail?

**Comment by MarcSabatella** [comment]
The parsing in edit() is just for the sake of the "spell check" so it is harmless - and maybe useful - if it fails there.  I am not comfortable with replacing on each keystroke - which is how it would with edit() - because I think it would be disorienting and possibly glitchy to have your text changing as you type.  Better to do it at the end, just as wedo with "b" and "#", I think. Do the replacement right the very top of Harmony::endEdit(), before even calling Text::endEdit(), since the latter has a bit of a ripple effect the links.  At least, that's what makes sense to me at first glance. Still would need more extensive testing with respect to parts and transposition.

**Comment by Jojo-Schmitz** [comment]
How to get at the string there? Got to be from `EditData& ed`, but how?

**Comment by Jojo-Schmitz** [comment]
Travis failed for reasons unrelated to this PR

**Comment by lasconic** [comment]
What's the status?

**Comment by Jojo-Schmitz** [comment]
"Work in Progress" or "Help needed" ;-) I can't get it to work

**Comment by MarcSabatella** [comment]
_username should indeed work.  It also works to user the accessor function hUserName() as you were originally.  Seems to work fine in 2.2, except as I mentioned, more testing woud be required.  As you observed, master has more severe issues with chord symbols, so I didn't even bother trying that.

**Comment by Jojo-Schmitz** [comment]
Ok, so I'll better switch to 2.2 for this

File: libmscore/harmony.cpp [hunk 1]
```diff
@@ -735,6 +735,14 @@ bool Harmony::edit(EditData& ed)
...
       {
+      _userName.replace("\u1d12b", "bb"); // double-flat
+      _userName.replace("\u226d",  "b");  // flat
+      _userName.replace("\u266f",  "#");  // sharp
+      _userName.replace("\u1d12a", "x");  // double-sharp
+      _userName.replace("\u0394",  "t");  // &Delta;
+      _userName.replace("\u00d0",  "o");  // &deg;
+      _userName.replace("\u00f8",  "0");  // &oslash;
+      _userName.replace("\u00d8",  "0");  // &Oslash;
       TextBase::endEdit(ed);
```
Comment (Jojo-Schmitz) [inline libmscore/harmony.cpp]: in 2.2 this is `Text::endEdit();`, guess that might be why the above replacements are not working in master?

Comment (Jojo-Schmitz) [reply libmscore/harmony.cpp]: No, it isn't that, stepping thru with a debugger shows that the replacement happens, but gets reverted by the `endEdit`

**Comment by Jojo-Schmitz** [comment]
Not different in 2.2, it still doesn't work there, the replacement does take place as seen in debugger, but doesn't take any effect or gets reverted by the `endEdit`

File: libmscore/harmony.cpp (hunk 1 above)
Comment (MarcSabatella) [reply libmscore/harmony.cpp]: Odd, it worked for me when I tried it.  But there is stuff going on with transposition and links that could be complicating this.  I'm not worried, this strikes me as extremely low priority.  The supported way of entering these symbols is easier.

Comment (lasconic) [reply libmscore/harmony.cpp]: I agree. The discussion here shows that it will be hard to add this in 2.2 since we don't know what we will break :)

Comment (Jojo-Schmitz) [reply libmscore/harmony.cpp]: Well, the problem is that it doesn't work at all. If it were, it'd be a simple text replacement before any interpretation and only the previously known characters would ever make it into the score file.

Comment (Jojo-Schmitz) [reply libmscore/harmony.cpp]: @MarcSabatella: I'd really like to see what you had working

**Comment by Jojo-Schmitz** [comment]
very strange, I did the very same stuff in 2.2 and it didn't work there either

**Comment by Jojo-Schmitz** [comment]
very strange, I did the very same stuff in 2.2 and it didn't work there either, see #3475

File: libmscore/harmony.cpp [hunk 2]
```diff
@@ -735,6 +735,14 @@ bool Harmony::edit(EditData& ed)
 
 void Harmony::endEdit(EditData& ed)
       {
+      _userName.replace("\u1d12b", "bb"); // double-flat
+      _userName.replace("\u226d",  "b");  // flat
```
Comment (Jojo-Schmitz) [inline libmscore/harmony.cpp]: stupid mistake, it should be `\u266d`

**Comment by dmitrio95** [comment]
@Jojo-Schmitz as far as I understand, this PR is ready to be merged, isn't it?

**Comment by Jojo-Schmitz** [comment]
Yes, I think so

**Pull request merged**